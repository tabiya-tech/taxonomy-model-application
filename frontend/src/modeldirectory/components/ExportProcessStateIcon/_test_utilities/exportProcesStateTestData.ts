import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ExportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllExportProcessStatePermutations = (): ModelInfoTypes.ExportProcessState[] => {
  // generate all permutations of ExportProcessState status, errored, exportErrors, exportWarnings
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ExportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((exportErrors) => {
        allBooleans.forEach((exportWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              exportErrors: exportErrors,
              exportWarnings: exportWarnings,
            },
            downloadUrl: "https://example.com",
            timestamp: new Date(),
          });
        });
      });
    });
  });
  return allPermutations;
};
