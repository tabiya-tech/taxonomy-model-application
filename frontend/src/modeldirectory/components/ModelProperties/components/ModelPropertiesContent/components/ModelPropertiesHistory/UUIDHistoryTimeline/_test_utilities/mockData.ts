import { ModelInfoTypes } from "src//modelInfo/modelInfoTypes";

export function getOneFakeUUIDHistoryDetail(i: number): ModelInfoTypes.ModelHistory {
  return {
    UUID: "e26aca13-aa00-4346-9e83-edaf59314378" + i,
    id: "5e12f537f079fbdcb4e61bf" + i,
    localeShortCode: "ZA",
    name: "Model" + i,
    version: "V1.1." + i,
  };
}

export function getFakeUUIDHistoryDetailsArray(length: number): ModelInfoTypes.ModelHistory[] {
  return Array.from({ length }, (_, i) => getOneFakeUUIDHistoryDetail(i));
}

export function getFakeUUIDHistoryDetailsArrayWithUnresolvedUUIDS(length: number): ModelInfoTypes.ModelHistory[] {
  return Array.from({ length }, (_, i) => {
    const details = getOneFakeUUIDHistoryDetail(i);
    details.id = null;
    details.name = null;
    details.version = null;
    details.localeShortCode = null;
    return details;
  });
}
