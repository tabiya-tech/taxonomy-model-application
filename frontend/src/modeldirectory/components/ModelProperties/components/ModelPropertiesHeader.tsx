import React from "react";
import { Box, useTheme } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import HeaderTitle from "src/theme/HeaderTitle/HeaderTitle";
import CustomIconButton from "src/theme/IconButton/CustomIconButton";

export interface ModelPropertiesHeaderProps {
  title: string;
  notifyOnClose: () => void;
}

const uniqueId = "a9f38045-7a8d-4226-88c6-8c9384516012";

export const DATA_TEST_ID = {
  MODEL_PROPERTIES_HEADER: `model-properties-header-${uniqueId}`,
  MODEL_PROPERTIES_HEADER_TITLE: `header-title-${uniqueId}`,
  MODEL_PROPERTIES_HEADER_CLOSE_BUTTON: `header-close-button-${uniqueId}`,
};

/**
 * ModelPropertiesHeader responsible for rendering the header of the model properties drawer
 * with a title and close button
 * @param props
 * @constructor
 */

const ModelPropertiesHeader: React.FC<ModelPropertiesHeaderProps> = (props: Readonly<ModelPropertiesHeaderProps>) => {
  const theme = useTheme();

  function handleOnClick() {
    try {
      props.notifyOnClose();
    } catch (error) {
      console.error(error);
    }
  }

  return (
    <Box
      display="flex"
      justifyContent="space-between"
      flexDirection="row"
      flexWrap="nowrap"
      alignItems="start"
      width="100%"
      gap={theme.spacing(theme.tabiyaSpacing.xl)}
      data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HEADER}
    >
      <HeaderTitle data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HEADER_TITLE}>{props.title}</HeaderTitle>
      <CustomIconButton
        onClick={handleOnClick}
        title="Close model properties"
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON}
      >
        <CloseIcon />
      </CustomIconButton>
    </Box>
  );
};

export default ModelPropertiesHeader;
