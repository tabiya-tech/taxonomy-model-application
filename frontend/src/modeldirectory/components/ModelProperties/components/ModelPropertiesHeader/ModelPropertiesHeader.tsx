import React from "react";
import { Box, Typography } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import HeaderTitle from "src/theme/HeaderTitle/HeaderTitle";
import CustomIconButton from "src/theme/IconButton/CustomIconButton";

export interface ModelPropertiesHeaderProps {
  name: string;
  notifyOnClose: () => void;
}

const uniqueId = "a9f38045-7a8d-4226-88c6-8c9384516012";

export const DATA_TEST_ID = {
  MODEL_PROPERTIES_HEADER: `model-properties-header-${uniqueId}`,
  MODEL_PROPERTIES_HEADER_TITLE: `header-title-${uniqueId}`,
  MODEL_PROPERTIES_MODEL_NAME: `model-name-${uniqueId}`,
  MODEL_PROPERTIES_HEADER_CLOSE_BUTTON: `header-close-button-${uniqueId}`,
};

/**
 * ModelPropertiesHeader responsible for rendering the header of the model properties drawer
 * with a title and close button
 * @param props
 * @constructor
 */

const ModelPropertiesHeader: React.FC<ModelPropertiesHeaderProps> = (props: Readonly<ModelPropertiesHeaderProps>) => {
  function handleOnClick() {
    try {
      props.notifyOnClose();
    } catch (error) {
      console.error(error);
    }
  }

  return (
    <Box display="flex" flexDirection="column" data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HEADER}>
      <Box display="flex" justifyContent="space-between" flexDirection="row" alignItems="start" width="100%">
        <Box flex={1} minWidth={0}>
          <HeaderTitle data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HEADER_TITLE}>Model Properties</HeaderTitle>
        </Box>
        <CustomIconButton
          onClick={handleOnClick}
          title="Close model properties"
          data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON}
        >
          <CloseIcon />
        </CustomIconButton>
      </Box>
      <Box
        flex={1}
        minWidth={0}
        sx={{
          display: "-webkit-box",
          WebkitLineClamp: "2",
          WebkitBoxOrient: "vertical",
          textOverflow: "ellipsis",
          wordWrap: "break-word",
          overflow: "hidden",
        }}
      >
        <Typography variant="subtitle1" data-testid={DATA_TEST_ID.MODEL_PROPERTIES_MODEL_NAME}>
          {props.name}
        </Typography>
      </Box>
    </Box>
  );
};

export default ModelPropertiesHeader;
