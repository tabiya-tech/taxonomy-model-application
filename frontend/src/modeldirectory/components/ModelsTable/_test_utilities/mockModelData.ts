import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import { ImportProcessStateEnums } from "api-specifications/importProcessState/enums";

import { v4 as uuidv4 } from "uuid";

import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { getMockId } from "src/_test_utilities/mockMongoId";
import { getRandomLorem, getRandomString, getTestString } from "src/_test_utilities/specialCharacters";
import { faker } from "@faker-js/faker";
import { CELL_MAX_LENGTH } from "src/modeldirectory/components/ModelsTable/ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  const model = getArrayOfFakeModels(1)[0];
  model.id = getMockId(id ?? 1);
  return model;
}

export function getArrayOfFakeModels(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location
          .countryCode("alpha-3")
          .substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description:
        i % 2 === 0
          ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH)
          : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getArrayOfFakeModelsMaxLength(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export const fakeModel: ModelInfoTypes.ModelInfo = {
  UUID: "8d914eab-6f7d-4183-accc-a09f99887b39",
  UUIDHistory: ["8d914eab-6f7d-4183-accc-a09f99887b39"],
  description: "aw j zt   agrkasl dy ogtimpsauwumu l utrovthao syertm beawpxluhyudgzbbm",
  id: "000000000000000000000001",
  importProcessState: {
    id: "000000000000000000000001",
    result: {
      errored: true,
      parsingErrors: true,
      parsingWarnings: true,
    },
    status: ImportProcessStateEnums.Status.PENDING,
    createdAt: new Date("2023-10-18T17:35:10.571Z"),
    updatedAt: new Date("2023-10-18T17:35:12.571Z"),
  },
  exportProcessState: [
    {
      downloadUrl: "https://www.example.com/",
      id: "000000000000000000000001",
      result: {
        errored: true,
        exportErrors: true,
        exportWarnings: true,
      },
      status: ExportProcessStateAPISpecs.Enums.Status.PENDING,
      timestamp: new Date("2023-10-18T17:35:10.571Z"),
      createdAt: new Date("2023-10-18T17:35:10.571Z"),
      updatedAt: new Date("2023-10-18T17:35:12.571Z"),
    },
  ],
  locale: {
    UUID: "7c4ea00b-fbea-4a31-9c04-009c813f5f34",
    name: "Palestine",
    shortCode: "ETH",
  },
  name: "1/1 - fysmros tcgcnjbbrrev",
  path: "https://unwelcome-editorial.net/",
  releaseNotes: "Perferendis modi impedit necessitatibus a",
  released: true,
  tabiyaPath: "https://silver-vulture.biz",
  createdAt: new Date("2023-10-18T17:35:10.571Z"),
  updatedAt: new Date("2022-10-18T17:35:10.571Z"),
  version: "0.2.2",
};

export function getRandomImportStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getRandomExportStatus(id: number) {
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getOneFakeExportProcessState(i: number): ModelInfoTypes.ExportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomExportStatus(i),
    result: {
      errored: i % 2 === 0,
      exportErrors: i % 2 === 0,
      exportWarnings: i % 2 === 0,
    },
    downloadUrl: faker.internet.url(),
    timestamp: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}

export function getOneFakeImportProcessState(i: number): ModelInfoTypes.ImportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomImportStatus(i),
    result: {
      errored: i % 2 === 0,
      parsingErrors: i % 2 === 0,
      parsingWarnings: i % 2 === 0,
    },
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}
