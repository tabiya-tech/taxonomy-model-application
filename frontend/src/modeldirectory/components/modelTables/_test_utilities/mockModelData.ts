import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import {v4 as uuidv4} from "uuid";

import {ModelInfoTypes} from "src/modelInfo/modelInfoTypes";
import {getMockId} from "src/_test_utilities/mockMongoId";
import {getRandomLorem, getRandomString, getTestString} from "src/_test_utilities/specialCharacters";
import {faker} from "@faker-js/faker";
import {CELL_MAX_LENGTH} from "../ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  return getArrayOfFakeModels(id ? id : 1)[0];
}

export function getArrayOfFakeModels(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: Array.from({length: 20}, (_, i) => faker.lorem.word()).join(" ").substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location.countryCode("alpha-3").substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: i % 2 === 0 ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH) : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      importProcessState: {
        id: getMockId(10000 + i),
        status: getRandomStatus(i),
        result: {
          errored: i % 2 === 0,
          parsingErrors: i % 2 === 0,
          parsingWarnings: i % 2 === 0,
        }
      }
    };
    models.push(model);
  }
  return models
}

export function getArrayOfFakeModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      importProcessState: {
        id: getMockId(10000 + i),
        status: getRandomStatus(i),
        result: {
          errored: i % 2 === 0,
          parsingErrors: i % 2 === 0,
          parsingWarnings: i % 2 === 0,
        }
      }
    };
    models.push(model);
  }
  return models
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      importProcessState: {
        id: getMockId(10000 + i),
        status: getRandomStatus(i),
        result: {
          errored: i % 2 === 0,
          parsingErrors: i % 2 === 0,
          parsingWarnings: i % 2 === 0,
        }
      }
    };
    models.push(model);
  }
  return models
}

export function getArrayOfFakeModelsForSorting(): ModelInfoTypes.ModelInfo[] {
  const baseModel: Omit<ModelInfoTypes.ModelInfo, 'name' | 'version' | 'createdAt' | 'updatedAt'> = {
    id: 'base',
    UUID: uuidv4(),
    previousUUID: uuidv4(),
    originUUID: uuidv4(),
    locale: {
      UUID: uuidv4(),
      name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      shortCode: faker.location.countryCode("alpha-3").substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
    },
    description: getRandomLorem(CELL_MAX_LENGTH / 2),
    released: true,
    releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
    path: faker.internet.url(),
    tabiyaPath: faker.internet.url(),
    importProcessState: {
      id: getMockId(10000),
      status: getRandomStatus(1),
      result: {
        errored: true,
        parsingErrors: true,
        parsingWarnings: true,
      }
    }
  };

  return [
    {
      ...baseModel,
      id: '1',
      name: 'ancient',
      version: '1.0.0',
      createdAt: new Date('2022-01-01T00:00:00.000Z'),
      updatedAt: new Date('2022-01-01T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '2',
      name: 'older',
      version: '1.1.0',
      createdAt: new Date('2022-01-02T00:00:00.000Z'),
      updatedAt: new Date('2022-01-02T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '3',
      name: 'old',
      version: '1.1.1',
      createdAt: new Date('2022-01-03T00:00:00.000Z'),
      updatedAt: new Date('2022-01-03T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '4',
      name: 'new',
      version: '1.1.2',
      createdAt: new Date('2022-01-04T00:00:00.000Z'),
      updatedAt: new Date('2022-01-04T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '5',
      name: 'newer',
      version: '1.1.3',
      createdAt: new Date('2022-01-05T00:00:00.000Z'),
      updatedAt: new Date('2022-01-05T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '6',
      name: 'newest',
      version: '1.2.2',
      createdAt: new Date('2022-01-06T00:00:00.000Z'),
      updatedAt: new Date('2022-01-06T00:00:00.000Z'),
    }
  ];
}


export function getRandomStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}
