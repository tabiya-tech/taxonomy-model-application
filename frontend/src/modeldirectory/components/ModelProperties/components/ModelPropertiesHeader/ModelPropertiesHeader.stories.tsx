import ModelPropertiesHeader from "./ModelPropertiesHeader";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof ModelPropertiesHeader> = {
  title: "ModelDirectory/ModelProperties/ModelPropertiesHeader",
  component: ModelPropertiesHeader,
  tags: ["autodocs"],
  argTypes: { notifyOnClose: { action: "notifyOnClose" } },
};

export default meta;

type Story = StoryObj<typeof ModelPropertiesHeader>;

export const Shown: Story = {
  args: {
    name: "Model Properties",
  },
};
