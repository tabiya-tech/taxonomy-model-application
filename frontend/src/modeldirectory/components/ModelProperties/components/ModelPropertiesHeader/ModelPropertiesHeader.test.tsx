// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen, within } from "src/_test_utilities/test-utils";
import ModelPropertiesHeader, { DATA_TEST_ID } from "./ModelPropertiesHeader";
import userEvent from "@testing-library/user-event";
import { fakeModel, getOneFakeModel } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";

import { ALL_USERS, authorizationTests } from "src/_test_utilities/authorizationTests";

describe("ModelPropertiesHeader", () => {
  beforeEach(() => {
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  describe(
    // eslint-disable-next-line jest/valid-describe-callback,jest/valid-title
    authorizationTests.defaultName,
    authorizationTests.callback({
      name: "ModelPropertiesHeader",
      Component: <ModelPropertiesHeader model={getOneFakeModel(1)} notifyOnClose={jest.fn()} />,
      roles: ALL_USERS,
      testIds: [
        DATA_TEST_ID.MODEL_PROPERTIES_HEADER,
        DATA_TEST_ID.MODEL_PROPERTIES_HEADER_TITLE,
        DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON,
        DATA_TEST_ID.MODEL_PROPERTIES_MODEL_NAME,
      ],
    })
  );

  test("should render model properties header component", () => {
    // GIVEN a model
    const givenModel = fakeModel;
    // AND a notifyOnClose callback function
    const givenNotifyOnCloseCallback = jest.fn();

    // WHEN ModelPropertiesHeader component is rendered
    render(<ModelPropertiesHeader model={givenModel} notifyOnClose={givenNotifyOnCloseCallback} />);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the header to be displayed
    const headerElement = screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER);
    expect(headerElement).toBeInTheDocument();
    // AND the title to be displayed in the header
    const titleElement = within(headerElement).getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER_TITLE);
    expect(titleElement).toBeInTheDocument();
    // AND the close icon button to be displayed in the header
    const closeButtonElement = within(headerElement).getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON);
    expect(closeButtonElement).toBeInTheDocument();
    // AND the model name to be displayed in the header and be properly formatted: "model name (locale short code)"
    const modelNameElement = within(headerElement).getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_MODEL_NAME);
    expect(modelNameElement).toBeInTheDocument();
    expect(modelNameElement).toHaveTextContent(`${givenModel.name} (${givenModel.locale.shortCode})`);
    // AND to match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER)).toMatchSnapshot();
  });

  test("should call notifyOnClose when the close icon button is clicked", async () => {
    // GIVEN notifyOnClose callback function
    const givenNotifyOnCloseCallback = jest.fn();
    // AND the ModelPropertiesHeader component is rendered
    render(<ModelPropertiesHeader model={getOneFakeModel(1)} notifyOnClose={givenNotifyOnCloseCallback} />);

    // WHEN the close icon button is clicked
    await userEvent.click(screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON));

    // THEN expect the notifyOnClose callback to be triggered
    expect(givenNotifyOnCloseCallback).toHaveBeenCalled();
  });

  test("should catch the error and log it when the parent's notifyOnClose throws an error", async () => {
    // GIVEN a notifyOnClose callback function that will throw an error
    const givenError = new Error("Error in notifyOnClose");
    const givenNotifyOnCloseCallback = jest.fn(() => {
      throw givenError;
    });
    // AND the ModelPropertiesHeader component is rendered with the given notifyOnClose callback
    render(<ModelPropertiesHeader model={getOneFakeModel(1)} notifyOnClose={givenNotifyOnCloseCallback} />);

    // WHEN the close icon button is clicked
    await userEvent.click(screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON));

    // THEN expect the given error to be logged
    expect(console.error).toHaveBeenCalledWith(givenError);
  });

  test("should catch the error and log it when the parent's notifyOnClose was not given", async () => {
    // GIVEN the ModelPropertiesHeader component is rendered without a notifyOnClose callback
    // @ts-ignore
    render(<ModelPropertiesHeader model={getOneFakeModel(1)} notifyOnClose={null} />);

    // WHEN the close icon button is clicked
    await userEvent.click(screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_HEADER_CLOSE_BUTTON));

    // THEN expect an error to be logged
    expect(console.error).toHaveBeenCalled();
  });
});
