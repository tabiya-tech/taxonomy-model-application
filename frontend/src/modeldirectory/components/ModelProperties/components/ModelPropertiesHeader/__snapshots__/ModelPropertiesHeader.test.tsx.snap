// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelPropertiesHeader should render model properties header component 1`] = `
<div
  class="MuiBox-root css-j7qwjs"
  data-testid="model-properties-header-a9f38045-7a8d-4226-88c6-8c9384516012"
>
  <div
    class="MuiBox-root css-rvqy99"
  >
    <div
      class="MuiBox-root css-1fjtzvx"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-2jbldj-MuiTypography-root"
        data-testid="header-title-a9f38045-7a8d-4226-88c6-8c9384516012"
      >
        Model Properties
      </h2>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1e844z7-MuiButtonBase-root-MuiIconButton-root"
      data-testid="header-close-button-a9f38045-7a8d-4226-88c6-8c9384516012"
      tabindex="0"
      title="Close model properties"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
        data-testid="CloseIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiBox-root css-1ygqla7"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 css-umxpai-MuiTypography-root"
      data-testid="model-name-a9f38045-7a8d-4226-88c6-8c9384516012"
    >
      Model Name
    </h6>
  </div>
</div>
`;
